pageserver/src/waldecoder.rs

```diff
@@ -184,16 +187,13 @@ impl WalStreamDecoder {
                     // XLOG_SWITCH records are special. If we see one, we need to skip
                     // to the next WAL segment.
                     if is_xlog_switch_record(&recordbuf) {
-                        trace!(
-                            "saw xlog switch record at {:X}/{:X}",
-                            (self.lsn >> 32),
-                            self.lsn & 0xffffffff
-                        );
-                        self.padlen = (WAL_SEGMENT_SIZE - (self.lsn % WAL_SEGMENT_SIZE)) as u32;
+                        trace!("saw xlog switch record at {}", self.lsn);
+                        self.padlen = (WAL_SEGMENT_SIZE - (self.lsn.0 % WAL_SEGMENT_SIZE)) as u32;
                     }
 
-                    if self.lsn % 8 != 0 {
-                        self.padlen = 8 - (self.lsn % 8) as u32;
+                    // FIXME: what does this code do?
+                    if self.lsn.0 % 8 != 0 {
+                        self.padlen = 8 - (self.lsn.0 % 8) as u32;
                     }
```

Comment:
Could someone educate me about what's happening here? It looks like:
- If "switch record" (what's that?), then compute padding to the end of the segment.
- If the current LSN isn't a multiple of 8, then compute padding to the next 8-byte boundary, overwriting the padding computed above (?!)

How much of this should be captured into `Lsn` member functions? We could add something like `Lsn::pad_to(chunk_size: u64)` that could be used for both.

Reply:
WAL in postgres is splitted into segments. Default size is 16Mb, but unlike XLOGBCLSZ is can be easily changed.
Switching XLOG means start writing to the next segment file.
Usually XLOG seitch record is not needed: WAL segments is filled till the end and then we swutch to next segment. But in some cases we may need to force switching to the new segment although there is still space in the current segment. It may be needed for WAL archiving.

WAL records are aligned on 8 byte boundary. Padding is used to enforce such alignment.
Actually the code fragment can be simplified as:
```
self.padlen  = (-self.lsn.0  & 7) as u32;
```

Reply:
If you want to see an xlog switch in action, you can force it with `select pg_switch_wal();`. Can become handy in testing. See https://www.postgresql.org/docs/10/functions-admin.html.

Reply:
Is it a bug that the code overwrites the "switch record" padding with the 8-byte wal record padding?